.github/actions/test/action.yml

```diff
@@ -149,7 +149,7 @@ runs:
       shell: bash
   
     - name: Upload Test Results
-      if: always()
+      if: '!cancelled()'
```

Comment:
I'd read this as a string, since it is a non-falsy value, treated as `true`. Then, this would not run on failure, only success.

```
Note that in conditionals, falsy values (false, 0, -0, "", '', null) are coerced to false and truthy (true and other non-falsy values) are coerced to true.
```
https://docs.github.com/en/actions/writing-workflows/choosing-what-your-workflow-does/evaluate-expressions-in-workflows-and-actions#literals

However, it looks like this is not the case, here the tests fail and the upload still is run:
https://github.com/EnricoMi/publish-unit-test-result-action/actions/runs/15241155700/job/42861390318?pr=669#step:3:841

Reply:
That's because you misinterpret how it works on GitHub and how YAML is parsed.
From YAML perspective `always()` is the same as the `String` `'always()`.
Just the same `${{ always() }}` is the same as the `String` `'${{ always() }}'`.
With `!canelled()` things are slightly different, because without the quotes it would be a YAML tag, so you need to explicitly use `'!cancelled()'`, or the full expression syntax `${{ !cancelled() }}` or even `'${{ !cancelled() }}'`.

All these are just Strings after the YAML parsing. How these Strings are then interpreted by GitHub Actions is a separate topic. And for `if` properties, the expression syntax is optional and the string is always interpreted as expression. That's why it works as intended. :-)